### ISSUE TITLE
[2.0 Bugfix] Bug in G27, M125 and M108

### ISSUE THREAD

**Issue by Alissonverd** [MarlinFirmware/Marlin#9254]
Hello.
I'm trying to use G27 or M125 associated with host pause feature. I need the hotend goes to a parking zone when I click in pause botton. This two commands have the same bug.

**Expected behaviour:**
When I click in pause botton, the print is paused and hotend go to X= 5 Y=5 and Z add +5.
When I click in resume botton, the printer will move to correct position XY and Z height.. and continue the print.

**Actual behaviour:**
Pause feature work fine. The hotend go to X5 Y5 and Z+5.
When click in resume, the printer make one layer in Z+5 and after move down to correct height.. So, after resume the print, we have one layer in the air.

**Steps to reproduce:**
1 - Pause the current print;
2 - Park hotend (G27 P2 or M125 L5)
3 - Resume the print.

===========
Other bug, command M108
I use the filament runout and send the M108 to avoid press botton on LCD. But this command dont make anything.

**Expected behaviour:**
After M600 trigger by filament sensor, I replace the filament and send M108 command.. So, the print continues.

**Actual behaviour:**
After M600 trigger by filament sensor, I replace the filament and send M108 command.. Nothing happens.. We still need press the LCD botton.
Pressing resume botton in printer host dont make effect too.

**Steps to reproduce:**
1 - M600 trigger by filament sensor or send it;
2 - Wait all procedures;
3 - Send M108;
4 - Nothing happens (printer dont continue)
5 - Press Resume in printer host (same as 4)

==================
**Edit.**
I'm doing some more tests.
After use M125 or G27 the filament runout dont trigger more.
**Steps to reproduce:**
1 - Use runout.
2 - Continue the print.
3 - Pause print and park hotend (M125 or G27)
4 - Resume print
5 - Try use runout again (notting happens)

**Comment by thinkyhead** [comment]
Hi! I'm just catching up on issues from the past month, and I see that this is closed without comment.

Did you solve the issue or find that it was already fixed in the current `bugfix-1.1.x` branch? Regarding the `M108` issue did you have `EMERGENCY_PARSER` enabled?

**Comment by Alissonverd** [comment]
@thinkyhead
The issue has not been resolved.
I closed because I thought they would not pay attention for perhaps not being the current focus.. 
And I was going to reopen the topic with a video showing the problems of the G27 and M125 commands

I use the last Marlin 2.0 bugfix with MKS GEN V1.4.. the printer host: OctoPrint 1.3.6

`M108` I will test again, I'm not sure that I tested with `EMERGENCY_PARSER`, but I think so. Anyway I will do it again to confirm in 20min.

-Edit:
Same again. `M108 with EMERGENCY_PARSER enabled`
When I pull out the filament (trigger the sensor), the printer does the standard procedures and the extruder removes the filament. The firmware send action::pause and OctoPrint stop the gcode file.
Now. When I send M108 in prompt.. Nothing happens, I dont have any msg about this command.
If I use the host button `resume,` to continue printing, nothing happens as well.. Continue with `Recv: echo:busy: paused for user`
_I have SDSUPPORT. I think I saw some limitations in code with this feature to pause and resume prints._

Now. I pressed the button (display) 1 time, waiting the temperature go to 215ºC and pressing again.. The result is.. The printer resume the print! The firmware send action::resume and Host restart send gcodes.

Why We need press 2 times?
Is much better if I press one time and firmware start the heatup and resume print. Maybe same with `M108` (when it works again). Sending the command, the firmware check filament sensor and start the hotend heatup, when temperature is ready, send the action::resume.

**Comment by Alissonverd** [comment]
@thinkyhead Could you reopen this issue?

**Comment by thinkyhead** [comment]
> Why We need press 2 times?

Bug in the LCD code? Sometimes improperly-coded features cause a click to get eaten. But there are also some little glitches in the LCD code that need to be worked out.

**Comment by Alissonverd** [comment]
@thinkyhead
Yes, using the LCD `REPRAP_DISCOUNT_SMART_CONTROLLER` after the hotend cooldown, we need press button 2 times to start print.
When we press one time, the hotend start heat up to a print temperature, and does not start printing automatically.
So, after heating done, we need other click to start print. If we not click, the hotend go to cooldown again.

I'm accustomed with the click, wait heatup and click again. But it would be much better just to click once and the whole process run automatically, and send action:resume to host.

**Comment by thinkyhead** [comment]
I'm sure I'll fix it next time I get into deep testing on the LCD controller. At the moment my personal focus is on clearing my unread issues, which now number close to 1000.

**Comment by Alissonverd** [comment]
Thank you! I'm available to do tests.

**Comment by Nutz95** [comment]
When i look at the code , in file **Gcode.cpp** for marlin version 2.0.x, it seems that EMERGENCY_PARSER is working reverse.
in the code to enable M108, M112 and M410 you need to DISABLED(EMERGENCY_PARSER)

here is what is written on the code: (inside file gcode.cpp around line 360)
#if **DISABLED**(EMERGENCY_PARSER)
        case 108: M108(); break;                                  // M108: Cancel Waiting
        case 112: M112(); break;                                  // M112: Emergency Stop
        case 410: M410(); break;                                  // M410: Quickstop - Abort all the planned moves.
      #endif

so to me M108, M112 and M410 cannot work if EMERGENCY_PARSER is Enabled.
I also saw somewhere else that the logic seems inverted and the use of DISABLED instead of ENABLED macro

Can someone confirm this? I've check in marlin 1.1.5 and 2.0.X from 08th of May 2018 and this part of the code seems to be here from quite some time.

**Comment by thinkyhead** [comment]
@Nutz95 — Please don't post comments regarding new issues on old issues. Post a new issue instead.

Besides, the code is correct.
See `MarlinSerial.cpp`, `emergency_parser.h`, and `emergency_parser.cpp`.

**Comment by mikeshub** [comment]
This bug still exists. I am going to mess with M125 instead of M600 as the runout script and see if this fixes it. Right now with M600 as the runout script and PARK_HEAD_ON_PAUSE defined it finishes the layer at the position traveled to with Z_raise in the nozzle park point. I'll report back what I find, but this is a pretty significant bug and this issue should be reopened and this bug should be fixed. M600 should undo the Z raise once the filament change is complete. 

EDIT: M125 for the runout script is even worse. It goes to the park position, returns to the last print position's X Y then gets stuck displaying "parking".

**Comment by ahmetcemturan** [comment]
It is the same with G27.  It will raise Z the specified amount but as long as there is no new Z coordinate in the following G-code it will remain in that raised Z position.   What we need is to return to the last position the G27 or M115 was executed.  Before the Gcode takes over.   otherwise it will just move to the next coordinate and as Gcde doesnt give Z coordinates every line anymore it keeps hovering above the print until the next Z-change.
